src/components/HistoryQuery.js

```diff
@@ -1,29 +1,88 @@
+/**
+ *  Copyright (c) Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the license found in the
+ *  LICENSE file in the root directory of this source tree.
+ */
+
 import React from 'react';
 import PropTypes from 'prop-types';
 
-const HistoryQuery = ({query, variables, operationName, onSelect}) => {
-  const onClick = () => {
-    onSelect(query, variables, operationName);
+export default class HistoryQuery extends React.Component {
+  static propTypes = {
+    query: PropTypes.string,
+    variables: PropTypes.string,
+    operationName: PropTypes.string,
+    favorite: PropTypes.bool,
+    favoriteSize: PropTypes.number,
+    onToggleFavorites: PropTypes.func,
+    onSelect: PropTypes.func,
   };
 
-  let displayName;
-  if (operationName) {
-    displayName = operationName;
-  } else {
-    displayName = query.split('\n')
-      .filter(line => line.indexOf('#') !== 0).join('');
+  constructor(props) {
+    super(props);
+    const favorite = this.props.favorite || false;
+    const starIcon = this.getStarIcon(favorite);
+    this.state = { favorite, starIcon };
+  }
+
+  render() {
+    const starStyles = {
+      float: 'right',
+    };
+    let displayName;
+    if (this.props.operationName) {
+      displayName = this.props.operationName;
+    } else {
+      displayName = this.props.query
+        .split('\n')
+        .filter(line => line.indexOf('#') !== 0)
+        .join('');
+    }
+    return (
+      <div>
+        <p onClick={this.handleClick.bind(this)}>
+          <span>
+            {displayName}
+          </span>
+          <span onClick={this.handleStarClick.bind(this)} style={starStyles}>
+            {this.state.starIcon}
+          </span>
+        </p>
+      </div>
+    );
   }
 
-  return (
-    <p onClick={onClick}>{displayName}</p>
-  );
-};
+  getStarIcon(isFavorite) {
+    return isFavorite ? '\u2605' : '\u2606';
+  }
 
-HistoryQuery.propTypes = {
-  query: PropTypes.string,
-  variables: PropTypes.string,
-  operationName: PropTypes.string,
-  onSelect: PropTypes.func,
-};
+  handleClick() {
+    this.props.onSelect(
+      this.props.query,
+      this.props.variables,
+      this.props.operationName,
+    );
+  }
 
-export default HistoryQuery;
+  handleStarClick(e) {
+    e.stopPropagation();
+    if (this.state.favorite === true) {
+      this.setState({
+        favorite: false,
+        starIcon: this.getStarIcon(false),
+      });
+    } else if (this.state.favorite === false) {
+      this.setState({
+        favorite: true,
+        starIcon: this.getStarIcon(true),
+      });
+    }
```

Comment:
Seems this whole `if` could be replaced with something like this:

```
this.setState({
  favorite: !this.state.favorite,
  starIcon: this.getStarIcon(!this.state.favorite),
});
```

or similar.